packages/fxa-payments-server/src/components/CouponForm/index.tsx

```diff
@@ -1,29 +1,64 @@
-import React, { useState, MouseEventHandler, FormEventHandler } from 'react';
-import { Localized } from '@fluent/react';
 import './index.scss';
+
+import { Localized } from '@fluent/react';
+import React, { FormEventHandler, MouseEventHandler, useState } from 'react';
+
+import { APIError, apiInvoicePreview } from '../../lib/apiClient';
 import { Coupon } from '../../lib/Coupon';
+import sentry from '../../lib/sentry';
+import { Plan } from '../../store/types';
+
+/*
+* Check if the coupon promotion code provided by the user is valid.
+* If it is valid, return the discounted amount in cents.
+*/
+const checkCouponCode = async (selectedPlan: Plan, couponCode: string) => {
+  const { plan_id: priceId } = selectedPlan;
+  try {
+    const { discount } = await apiInvoicePreview({
+      priceId,
+      promotionCode: couponCode
+    });
+
+    if(discount)
+      return discount.amount;
+
+    throw new Error('No discount for coupon');
+  } catch (err) {
+    if (err instanceof APIError)
+      sentry.captureException(err);
+    throw err;
+  }
+}
 
 type CouponFormProps = {
+  selectedPlan: Plan;
   coupon?: Coupon;
   setCoupon: (coupon: Coupon | undefined) => void;
 };
 
-export const CouponForm = ({ coupon, setCoupon }: CouponFormProps) => {
+export const CouponForm = ({ selectedPlan, coupon, setCoupon }: CouponFormProps) => {
   const [hasCoupon, setHasCoupon] = useState(coupon ? true : false);
   const [couponCode, setCouponCode] = useState(coupon ? 'test' : '');
   const [error, setError] = useState(false);
+  const [loading, setLoading] = useState(false);
 
-  const onSubmit: FormEventHandler = (event) => {
+  const onSubmit: FormEventHandler = async (event) => {
     event.preventDefault();
     event.stopPropagation();
-    if (couponCode === 'test') {
+    try {
+      setLoading(true);
+      const amount = await checkCouponCode(selectedPlan, couponCode);
       setHasCoupon(true);
       setCoupon({
-        amount: 200,
-        couponCode: '',
+        couponCode: couponCode,
+        amount,
       });
-    } else {
+    } catch (err) {
+      setCoupon(undefined);
       setError(true);
```

Comment:
If I recall from the other PR, setting error to true here would render the error message about the promotion code being invalid or expired.  But the `err` caught here could be an API error, and in that case, we'd be misinforming the user.

Reply:
I just tested this, and Barry is correct. It's kind of the difference between a 400 and a 500 server response.

I don't think it needs to block this PR, but perhaps it's worth discussing in the next design check-in if we want to add at least one other generic error message here in the event there was a server error (i.e. nothing is wrong with the code the user entered, and nothing they can do, perhaps other than wait for us to fix it, will change the result). Can you add a topic to the agenda to this effect Reino?

At the very least for this PR, perhaps we can capture the error in Sentry if it's an API error?

Reply:
I thought it best to send the error to Sentry in the `checkCouponCode` (`checkPromotionCode`) function.

I did update the logic a bit to send a different error message. I'll create a separate issue to properly implement this, but at least we'll have something for now.